tests/Tests/Unit/SerializerTest.php

```diff
@@ -291,4 +291,50 @@ public function testSpecialEncodingDecodingByFieldType()
         $this->assertEqualsWithDelta($data['blue'], $array['blue'], 0.0000001);
         $this->assertEqualsWithDelta($data['alpha']['value'], $array['alpha']['value'], 0.0000001);
     }
+
+    /**
+     * @dataProvider customEncoderProvider
+     */
+    public function testCustomEncoderForEncodeMessage($customEncoder, $protoObj, $expectedData)
+    {
+        $serializer = new Serializer([], [], [], [], $customEncoder);
+
+        $data = $serializer->encodeMessage($protoObj);
+        array_walk($expectedData, function ($val, $key, $expectedData) {
+            $this->assertEquals($val, $expectedData[$key]);
+        }, $expectedData);
+    }
+
+    public function customEncoderProvider()
+    {
+        $original = [
+            'red' => 50.0,
+            'blue' => 50.0
+        ];
+
+        $modified = [
+            'red' => 100,
+            'blue' => 100
+        ];
+
+        return  [
+            [
+                [
+                    Color::class => function ($message) use($modified) {
+                        return $modified;
+                    }
+                ],
+                new Color($original),
+                $modified
+            ],
+            [
+                // When no custom encoder is supplied, the encodeMessage will return the data
```

Comment:
another nit, but in addition to supplying NO encoder, I'd like to see a test where an encoder is supplied for a _different_ message class, to illustrate that only matching classes are affected.

Reply:
Yeah, good idea.